File: Leetcode/dfs_recursion_backtracking/654_MaximumBinaryTree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode constructMaximumBinaryTree(int[] nums) {
        if(nums.length==0)
            return null;
        return dfs(nums, 0, nums.length-1);
    }
        private TreeNode dfs(int[] nums, int lower,
                             int higher){
            if(lower>higher){
                return null;
            }
            int maxValue = nums[lower], maxIndex = lower;
            for(int i = lower; i<=higher; ++i){
                if(nums[i]>maxValue){
                    maxValue = nums[i];
                    maxIndex = i;
                }
            }
            TreeNode root = new TreeNode(maxValue);
            root.left = dfs(nums, lower, maxIndex-1);
            root.right = dfs(nums, maxIndex+1, higher);
            return root;
        }
}

File: Leetcode/dfs_recursion_backtracking/654_MaximumBinaryTree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        if not nums:
            return None
        maxValue = nums[0]
        maxIndex = 0;
        for i, num in enumerate(nums):
            if num > maxValue:
                maxValue = num
                maxIndex = i
        root = TreeNode(maxValue)
        root.left = self.constructMaximumBinaryTree(nums[:maxIndex])
        root.right = self.constructMaximumBinaryTree(nums[maxIndex+1:])
        return root

